storage/pruning/fullHistoryPruningStorer.go

```diff
@@ -0,0 +1,133 @@
+// NewFullHistoryPruningStorer will return a new instance of PruningStorer without sharded directories' naming scheme
+func NewFullHistoryPruningStorer(args *FullHistoryStorerArgs) (*FullHistoryPruningStorer, error) {
+	return initFullHistoryPruningStorer(args, "")
+}
+
+// NewShardedFullHistoryPruningStorer will return a new instance of PruningStorer with sharded directories' naming scheme
+func NewShardedFullHistoryPruningStorer(
+	args *FullHistoryStorerArgs,
+	shardID uint32,
+) (*FullHistoryPruningStorer, error) {
+	shardStr := fmt.Sprintf("%d", shardID)
+	return initFullHistoryPruningStorer(args, shardStr)
+}
+
+func initFullHistoryPruningStorer(args *FullHistoryStorerArgs, shardId string) (*FullHistoryPruningStorer, error) {
+	ps, err := initPruningStorer(args.StorerArgs, shardId)
+	if err != nil {
+		return nil, err
+	}
+
+	if args.NumOfOldActivePersisters < 1 || args.NumOfOldActivePersisters > math.MaxInt32 {
+		return nil, storage.ErrInvalidNumberOfOldPersisters
+	}
+
+	oldEpochsActivePersisters, err := lrucache.NewCacheWithEviction(int(args.NumOfOldActivePersisters), onEvicted)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &FullHistoryPruningStorer{
+		PruningStorer:             ps,
+		args:                      args.StorerArgs,
+		shardId:                   shardId,
+		oldEpochsActivePersisters: oldEpochsActivePersisters,
+	}, nil
+}
+
+func onEvicted(key interface{}, value interface{}) {
+	pd, ok := value.(*persisterData)
+	if ok {
+		err := pd.Close()
+		if err != nil {
+			log.Warn("initFullHistoryPruningStorer - onEvicted", "key", key, "err", err.Error())
+		}
+	}
+}
+
+// GetFromEpoch will search a key only in the persister for the given epoch
+func (fhps *FullHistoryPruningStorer) GetFromEpoch(key []byte, epoch uint32) ([]byte, error) {
+	data, err := fhps.searchInEpoch(key, epoch)
+
+	if err == nil && data != nil {
+		return data, nil
+	}
+
+	return fhps.searchInEpoch(key, epoch+1)
+}
+
+func (fhps *FullHistoryPruningStorer) searchInEpoch(key []byte, epoch uint32) ([]byte, error) {
+	fhps.lock.RLock()
+	oldestEpochInCurrentSetting := fhps.activePersisters[len(fhps.activePersisters)-1].epoch
+	fhps.lock.RUnlock()
+
+	isActiveEpoch := epoch > oldestEpochInCurrentSetting-fhps.numOfActivePersisters
+	if isActiveEpoch {
+		return fhps.PruningStorer.Get(key)
+	}
+
+	data, err := fhps.getFromOldEpoch(key, epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func (fhps *FullHistoryPruningStorer) getFromOldEpoch(key []byte, epoch uint32) ([]byte, error) {
+	epochString := fmt.Sprintf("%d", epoch)
+
+	fhps.lock.RLock()
+	pdata, exists := fhps.oldEpochsActivePersisters.Get([]byte(epochString))
+	fhps.lock.RUnlock()
+	if !exists {
+		newPdata, errPersisterData := createPersisterDataForEpoch(fhps.args, epoch, fhps.shardId)
+		if errPersisterData != nil {
+			return nil, errPersisterData
+		}
+
+		fhps.oldEpochsActivePersisters.Put([]byte(epochString), newPdata, 0)
+		fhps.persistersMapByEpoch[epoch] = newPdata
```

Comment:
shouldn't this be protected by mutex?
I think get-and-open-if-missing should be done in a single Lock-Unlock area as to avoid overwriting persisters that will never get closed (L101-L119). Performance is still desired but not mandatory here.